File: architecture-skills/object-oriented-programming-paradigm/examples/interface-segregation/interface-segregation.js
// Interface Segregation Principle in JavaScript.
//
// A fat officeMachine duck type forces SimplePrinter to stub scanDocument
// and faxDocument with errors. The fix splits printer and scanner shapes
// so each client depends only on the methods it calls.

class NotSupportedError extends Error {
  constructor(feature) {
    super(feature);
    this.name = 'NotSupported';
  }
}

// --- Violation: one fat shape every device must implement ---

class SimplePrinter {
  printDocument(path) {
    console.log(`print ${path}`);
  }

  scanDocument(path) {
    throw new NotSupportedError('scan');
  }

  faxDocument(path) {
    throw new NotSupportedError('fax');
  }
}

function runOfficeJob(machine, path) {
  machine.printDocument(path);
  for (const [action, fn] of [
    ['scan', () => machine.scanDocument(path)],
    ['fax', () => machine.faxDocument(path)],
  ]) {
    try {
      fn();
    } catch (e) {
      if (e instanceof NotSupportedError) {
        console.log(`${action} attempt: NotSupported`);
      } else {
        throw e;
      }
    }
  }
}

// --- Fix: small client-specific shapes ---

class BasicPrinter {
  printDocument(path) {
    console.log(`print ${path}`);
  }
}

class FlatbedScanner {
  scanDocument(path) {
    console.log(`scan ${path}`);
  }
}

class MultiFunctionDevice {
  printDocument(path) {
    console.log(`print ${path}`);
  }

  scanDocument(path) {
    console.log(`scan ${path}`);
  }
}

function printJob(printer, path) {
  printer.printDocument(path);
}

function scanJob(scanner, path) {
  scanner.scanDocument(path);
}

function main() {
  console.log('-- Violation: SimplePrinter forced to stub scan/fax --');
  runOfficeJob(new SimplePrinter(), 'document.pdf');

  console.log('\n-- Fix: clients depend only on what they need --');
  printJob(new BasicPrinter(), 'document.pdf');
  scanJob(new FlatbedScanner(), 'page.tif');
  printJob(new MultiFunctionDevice(), 'report.pdf');
}

main();

File: architecture-skills/object-oriented-programming-paradigm/examples/interface-segregation/interface-segregation.swift
// Interface Segregation Principle in Swift.
//
// A fat OfficeMachine protocol forces SimplePrinter to stub scanDocument
// and faxDocument with errors. The fix splits Printer and Scanner so
// each client depends only on the methods it calls.

import Foundation

enum OfficeError: Error, CustomStringConvertible {
    case notSupported(String)

    var description: String { "not supported" }
}

// --- Violation: one fat protocol every device must implement ---

protocol OfficeMachine {
    func printDocument(_ path: String)
    func scanDocument(_ path: String) throws
    func faxDocument(_ path: String) throws
}

final class SimplePrinter: OfficeMachine {
    func printDocument(_ path: String) {
        print("print \(path)")
    }

    func scanDocument(_ path: String) throws {
        throw OfficeError.notSupported("scan")
    }

    func faxDocument(_ path: String) throws {
        throw OfficeError.notSupported("fax")
    }
}

func runOfficeJob(_ machine: OfficeMachine, path: String) {
    machine.printDocument(path)
    for (action, fn) in [
        ("scan", { try machine.scanDocument(path) }),
        ("fax", { try machine.faxDocument(path) }),
    ] {
        do {
            try fn()
        } catch is OfficeError {
            print("\(action) attempt: NotSupported")
        } catch {
            fatalError("\(error)")
        }
    }
}

// --- Fix: small client-specific protocols ---

protocol Printer {
    func printDocument(_ path: String)
}

protocol Scanner {
    func scanDocument(_ path: String)
}

final class BasicPrinter: Printer {
    func printDocument(_ path: String) { print("print \(path)") }
}

final class FlatbedScanner: Scanner {
    func scanDocument(_ path: String) { print("scan \(path)") }
}

final class MultiFunctionDevice: Printer, Scanner {
    func printDocument(_ path: String) { print("print \(path)") }
    func scanDocument(_ path: String) { print("scan \(path)") }
}

func printJob(_ printer: Printer, path: String) {
    printer.printDocument(path)
}

func scanJob(_ scanner: Scanner, path: String) {
    scanner.scanDocument(path)
}

print("-- Violation: SimplePrinter forced to stub scan/fax --")
runOfficeJob(SimplePrinter(), path: "document.pdf")

print("\n-- Fix: clients depend only on what they need --")
printJob(BasicPrinter(), path: "document.pdf")
scanJob(FlatbedScanner(), path: "page.tif")
printJob(MultiFunctionDevice(), path: "report.pdf")
